### PR TITLE
pull request to fix #23360

### LINKED ISSUE #23360: \OC\Files hooks don't fire, for folder which only contains folders
I writed an app that changes file ownership on the filesystem to actual ldap user who uploaded or created new files and folders. The reason why I done that is wanted to use linux filesystem quota for some reasons. 

To achieve this I used files hooks explained in the <a href="https://doc.owncloud.org/server/8.2/developer_manual/app/hooks.html?highlight=hook">Developper Manual 8.2</a>

In the app, I used postWrite and postCreate hooks to listen write and create events.

In my tests, I realized that if I uploaded a folder which contains only folders, but inner folders may have files and folder, hooks does not fire for only mentioned type folders.

I would see this as bug. Any help is very welcome...

### Steps to reproduce
1. Create a new folder
2. Create new subfolder inside the created folder in first step
3. Create a file in inside the created folder in second step
4. Drag and drop created folder in first step to web client
5. Listen postWrite and postCreate hooks

### Expected behaviour
Hooks should fire for all elements 

### Actual behaviour
Hooks does not fire for created folder in first step which contains only folder, for other elements it works fine.

### Server configuration
**Operating system**: RHEL 7.2
**Web server:** apache
**Database:** postgresql
**PHP version:** 5.4
**ownCloud version:** (see ownCloud admin page) 8.2.2
**Updated from an older ownCloud or fresh install:** fresh install
**Where did you install ownCloud from:** OwnCloud Rhel repos
**Are you using an external user-backend, if yes which one:** ActiveDirectory

### PR DESCRIPTION
**Pull request by karakayasemi**
fromTmpFile function, usual mkdir call is only working for file's parent directory. Does not care upper parent folders. I added a recursive function that creates parent non-existing folders with usual mkdir.

Fix #23360

### CODE REVIEW

File: lib/private/files/view.php [hunk 1]
```diff
@@ -2051,4 +2054,22 @@ public function getUidAndFilename($filename) {
 		}
 		return [$uid, $filename];
 	}
+	
+	/**
+	 * Creates parent non-existing folders
+	 * @param $filePath
+	 * @return bool
+	 */
+	public function createParentDirectories($filePath)
```
Comment (LukasReschke) [inline lib/private/files/view.php]: Should not be public

File: lib/private/files/view.php [hunk 2]
```diff
@@ -2051,4 +2054,22 @@ public function getUidAndFilename($filename) {
 		}
 		return [$uid, $filename];
 	}
+	
+	/**
+	 * Creates parent non-existing folders
+	 * @param $filePath
```
Comment (LukasReschke) [inline lib/private/files/view.php]: Missing type

File: lib/private/files/view.php [hunk 3]
```diff
@@ -2051,4 +2054,22 @@ public function getUidAndFilename($filename) {
 		}
 		return [$uid, $filename];
 	}
+	
+	/**
+	 * Creates parent non-existing folders
```
Comment (LukasReschke) [inline lib/private/files/view.php]: Add a empty new line after this

File: lib/private/files/view.php [hunk 4]
```diff
@@ -2051,4 +2054,22 @@ public function getUidAndFilename($filename) {
...
+	 * @param $filePath
+	 * @return bool
+	 */
+	public function createParentDirectories($filePath)
+	{
```
Comment (LukasReschke) [inline lib/private/files/view.php]: Our coding guidelines require curly braces to be on the same line as the function definition

File: lib/private/files/view.php [hunk 5]
```diff
@@ -2051,4 +2054,22 @@ public function getUidAndFilename($filename) {
...
+		$parentDirectory = dirname($filePath);
+		while(!$this->file_exists($parentDirectory)) {
+			$result=$this->createParentDirectories($parentDirectory);
+			if($result == false) {
```
Comment (LukasReschke) [inline lib/private/files/view.php]: `===`

File: lib/private/files/view.php [hunk 6]
```diff
@@ -2051,4 +2054,22 @@ public function getUidAndFilename($filename) {
...
+	/**
+	 * Creates parent non-existing folders
+	 * @param $filePath
+	 * @return bool
+	 */
+	public function createParentDirectories($filePath)
+	{
+		$parentDirectory = dirname($filePath);
+		while(!$this->file_exists($parentDirectory)) {
+			$result=$this->createParentDirectories($parentDirectory);
```
Comment (LukasReschke) [inline lib/private/files/view.php]: Space before and after `=`

**Comment by LukasReschke** [comment]
Can you add an unit test as well?

cc @icewind1991 What do you think?

**Comment by karakayasemi** [comment]
I hereby confirm that this contribution is MIT licensed.

**Comment by PVince81** [comment]
@karakayasemi forgive the bot, sometimes it's a bit clunky

**Comment by PVince81** [comment]
@karakayasemi thanks a lot. Let us know if you need help for writing unit tests.
See https://doc.owncloud.org/server/9.0/developer_manual/core/unit-testing.html#running-unit-tests-for-the-owncloud-core-project

Best start is to first run the test with `./autotest.sh sqlite tests/lib/files/view.php` and then later add a test function there.

**Comment by DeepDiver1975** [comment]
Commits shall be squashed - THX

**Comment by PVince81** [comment]
But maybe first have the unit tests and then squash later.

**Comment by icewind1991** [comment]
hmm, `mkdir` should already be recursive on the storage level

**Comment by karakayasemi** [comment]
on the storage level no hooks are triggering. In this case, hooks just triggers for file's parent folder on files\view basicOperation function before going to storage layer.

**Comment by PVince81** [comment]
For the test, what you could do is call `tmpFile` with a folder path, then assert that the folder was created properly.

**Comment by icewind1991** [comment]
if it's just a problem about hooks not being fired for implicitly created parent folders that that should be fixed instead of moving the parent folder creation logic to the View

**Comment by karakayasemi** [comment]
Maybe we can change file_put_contents function hook emitting logic instead of folder creation logic. In file_put_contents just one hook emitting for file's parent directory. Does not care upper parent folders. We can recursively emit hook signal for all upper parent folders

**Comment by karakayasemi** [comment]
Sorry folders are creating in basicOperation functions we can change hook emitting logic in basicOperation function.

**Comment by karakayasemi** [comment]
In storage mkdir implementation was made for recursive mkdir. But no hooks are emitting in mkdir. What do u think about change hook emitting logic in basicOperation function instead of moving the parent folder creation logic to the View? Or should we emit hook on storage level?

**Comment by PVince81** [comment]
@icewind1991 the reasoning in this PR sounds valid to me.
The thing about hooks only firing for the first parent is for different operations.

In this very specific use case, the series of folders gets created only because someone used "folder upload" in the web UI, so all parents need to be created before the files can be uploaded into them.

The creation of these folders need to somehow trigger a hook, but which one ?
The easiest approach here is to let it trigger the hook for each creation.

**Comment by PVince81** [comment]
@icewind1991 any update ? Are you still against this approach ?

**Comment by PVince81** [comment]
@owncloud/filesystem

**Comment by icewind1991** [comment]
This is not fully optimal but I'm fine with getting this merged.

Needs a rebase though

**Comment by PVince81** [comment]
Rebased and resent as https://github.com/owncloud/core/pull/25208

**Pull request closed without merging**